x-pack/plugins/security_solution/server/lib/entity_analytics/entity_store/update_entity_store.ts

```diff
@@ -0,0 +1,643 @@
+    }
+  }
+
+  return hits
+    .filter((hit) => hit._source)
+    .map((hit) => {
+      const src = hit._source as CompositeHit;
+      return {
+        id: hit._id,
+        '@timestamp': src['@timestamp'],
+        type: src.entity.type,
+        host: src.host,
+        ip_history: src.entity.ip_history,
+        first_doc_timestamp: src.entity.first_doc_timestamp,
+        last_doc_timestamp: src.entity.last_doc_timestamp,
+        latest_os_timestamp: src.entity.latest_os_timestamp,
+        latest_os: src.entity.latest_os,
+      };
+    });
+}
+
+async function getNextAssetCriticalities({
+  assetCriticalityService,
+  fromTimestamp,
+  lastProcessedId,
+}: {
+  assetCriticalityService: AssetCriticalityService;
+  fromTimestamp?: string;
+  lastProcessedId?: string;
+}): Promise<AssetCriticalityRecord[]> {
+  if (!fromTimestamp) {
+    return [];
+  }
+  const criticalities = await assetCriticalityService.getCriticalitiesFromTimestamp(
+    fromTimestamp,
+    MAX_CRITICALITY_SIZE
+  );
+
+  if (!lastProcessedId || !criticalities.length) {
+    return criticalities;
+  }
+
+  const lastProcessedIndex = criticalities.findIndex((criticality) => {
+    return (
+      AssetCriticalityDataClient.createIdFromRecord(criticality) === lastProcessedId &&
+      criticality['@timestamp'] === fromTimestamp
+    );
+  });
+
+  if (lastProcessedIndex === -1) {
+    return criticalities;
+  }
+
+  return criticalities.slice(lastProcessedIndex + 1);
+}
+
+async function getNextRiskScores({
+  riskScoreDataClient,
+  fromTimestamp,
+  lastProcessed,
+}: {
+  riskScoreDataClient: RiskScoreDataClient;
+  fromTimestamp?: string;
+  lastProcessed?: {
+    id_field: string;
+    id_value: string;
+  };
+}): Promise<EcsRiskScore[]> {
+  if (!fromTimestamp) {
+    return [];
+  }
+  const riskScores = await riskScoreDataClient.getRiskScoresFromTimestamp({
+    from: fromTimestamp,
+    size: MAX_CRITICALITY_SIZE,
+    namespace: 'default',
+  });
+
+  if (!lastProcessed || !riskScores.length) {
+    return riskScores;
+  }
+
+  const lastProcessedIndex = riskScores.findIndex((riskScore) => {
+    return (
+      riskScore.host?.risk.id_value === lastProcessed.id_value &&
+      riskScore.host?.risk.id_field === lastProcessed.id_field &&
+      riskScore['@timestamp'] === fromTimestamp
+    );
+  });
+
+  if (lastProcessedIndex === -1) {
+    return riskScores;
+  }
+
+  return riskScores.slice(lastProcessedIndex + 1);
+}
+
+function compareTimestamps(a: string, b: string) {
+  return moment(a).diff(moment(b));
+}
+
+function isTimestampBefore(a: string, b: string) {
+  return compareTimestamps(a, b) < 0;
+}
+
+function isTimestampAfter(a: string, b: string) {
+  return compareTimestamps(a, b) > 0;
+}
+
+function groupAndCombineCompositesByHostName(
+  composites: CompositeDocument[]
+): Record<string, CompositeDocument> {
+  const compositesGroupedByHostName = _.groupBy(composites, (composite) => composite.host.name);
```

Comment:
Because the history is stored on arrays like here:
```
ip_history.ip: [1.1.1.1, 2.2.2.2]
ip_history.timestamp: [t1, t2]
```
We can't easily implement it on ES using `terms` or `top_n` aggregations. We would have to write encrypted aggregations.